File: 3 - Longest Substring Without Repeating Characters.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // Time Complexity: O(N) since with sliding window you'll process each character at
        //                       most twice
        // Space Complexity: O(N) because we use a hashset to store unique characters
        unordered_set<char> chars;
        int l = 0;
        int longest = 0;
        for (int r = 0; r < s.length(); r++)
        {
            while (chars.find(s[r]) != chars.end())
            {
                // found duplicate so remove first instance
                chars.erase(s[l]);
                l++;
            }
            chars.insert(s[r]);
            longest = max(longest, r - l + 1);
        }
        return longest;
    }
};

int main() {
    Solution sol;
    string s;

    cout << "Enter a string to test longest substring without repeating characters:\n";
    cin >> s;   // <-- no getline, reads a single-token string

    int result = sol.lengthOfLongestSubstring(s);

    cout << "Input String: " << s << "\n";
    cout << "Longest Substring Length: " << result << "\n";
    return 0;
}

File: solutions/3 - Longest Substring Without Repeating Characters.java
import java.util.HashMap;
import java.util.Scanner;
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // Time Complexity: O(N) because we use sliding window.
        //                  HashMap operations are O(1).
        // Space Complexity: O(1) because bounded by number of unique chars.
        HashMap<Character, Integer> set = new HashMap<>();
        int ans = 0;
        int j = 0;
        for (int i = 0; i < s.length(); i++) {
            Character curr = s.charAt(i);
            if (set.containsKey(s.charAt(i))) {
                j = Math.max(set.get(curr) + 1, j);
            }
            ans = Math.max(ans, i - j + 1);
            set.put(curr, i);
            // System.out.println(j + " " + i);
        }
        return ans;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Solution sol = new Solution();

        System.out.println("Enter a string to test longest substring without repeating characters:");
        String input = sc.nextLine();

        int result = sol.lengthOfLongestSubstring(input);
        System.out.println("Input String: " + input);
        System.out.println("Longest Substring Length: " + result);
    }
}